sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala

```diff
@@ -49,37 +73,120 @@ class PythonMicroBatchStream(
   // from python to JVM.
   private var cachedInputPartition: Option[(String, String, PythonStreamingInputPartition)] = None
 
+  // Store the latest available offset for reporting
+  private var latestAvailableOffset: Option[PythonStreamingSourceOffset] = None
+
   private val runner: PythonStreamingSourceRunner =
     new PythonStreamingSourceRunner(createDataSourceFunc, outputSchema)
   runner.init()
 
   override def initialOffset(): Offset = PythonStreamingSourceOffset(runner.initialOffset())
 
-  override def latestOffset(): Offset = PythonStreamingSourceOffset(runner.latestOffset())
+  /**
+   * Returns the default read limit based on configured options. Supports: maxRecordsPerBatch,
+   * maxFilesPerBatch, maxBytesPerBatch. Falls back to allAvailable if no valid limit is
+   * configured.
+   *
+   * @since 4.2.0
+   */
+  override def getDefaultReadLimit: ReadLimit = {
+    import scala.util.Try
+
+    def parseLong(key: String): Long = {
+      Try(options.get(key).toLong).toOption.filter(_ > 0).getOrElse {
+        throw new IllegalArgumentException(
+          s"Invalid value '${options.get(key)}' for option '$key', must be a positive integer")
+      }
+    }
+
+    def parseInt(key: String): Int = {
+      Try(options.get(key).toInt).toOption.filter(_ > 0).getOrElse {
+        throw new IllegalArgumentException(
+          s"Invalid value '${options.get(key)}' for option '$key', must be a positive integer")
+      }
+    }
+
+    if (options.containsKey("maxRecordsPerBatch")) {
```

Comment:
It'd be preferred to use constant fields for constant strings - let's put constants in companion object.